js/index.d.ts

```diff
@@ -0,0 +1,122 @@
+/// <reference lib="es2020" />
+
+declare interface MimeType {
+    mime: string;
+}
+
+export interface Resource {
+    name: string;
+    aliases: string[];
+    kind: MimeType | "template";
+    content: string;
+    dependencies?: string[];
+    permission?: number;
+}
+
+export interface FilterListMetadata {
+    homepage?: string;
+    title?: string;
+    expires?: number;
+    redirect?: string;
+}
+
+export interface ParseOptions {
+    format: typeof FilterFormat;
+    rule_types: typeof RuleTypes;
+}
+
+declare interface BlockerResult {
+    matched: boolean,
+    important: boolean,
+    redirect?: string,
+    rewritten_url?: string,
+    exception?: string,
+    filter?: string,
+}
+
+export class Engine {
+    constructor(rules: FilterSet, debug: boolean);
+    addResource(resource: Resource): boolean;
+    check(url: string, source_url: string, request_type: string, debug?: false): boolean;
+    check(url: string, source_url: string, request_type: string, debug: true): BlockerResult;
+    clearTags(): null;
+    deserialize(serialized_handle: ArrayBuffer): null;
+    enableTag(tag: string): null;
+    getResource(name: string): Resource;
+    serializeCompressed(): ArrayBuffer;
+    serializeRaw(): ArrayBuffer;
+    tagExists(tag: string): boolean;
+    useResources(resources: Resource[]): null;
+}
+
+type CbType =
+    'block' |
+    'block-cookies' |
+    'css-display-none' |
+    'ignore-previous-rules' |
+    'make-https';
+
+declare interface CbAction {
+    type: CbType,
+    selector?: string,
+}
+
+type CbResourceType =
+    'document' |
+    'image' |
+    'style-sheet' |
+    'script' |
+    'font' |
+    'raw' |
+    'svg-document' |
+    'media' |
+    'popup';
+
+type CbLoadType =
+    'first-party' |
+    'third-party';
+
+declare interface CbTrigger {
+    'url-filter': string,
+    'url-filter-is-case-sensitive'?: boolean,
+    'if-domain'?: string[],
+    'unless-domain'?: string[],
+    'resource-type'?: Record<string, CbResourceType>,
+    'load-type'?: CbLoadType[],
+    'if-top-url'?: string[],
+    'unless-top-url'?: string[],
+}
+
+declare interface CbRule {
+    action: CbAction,
+    trigger: CbTrigger,
+}
+
+declare interface ContentBlockingConversionResult {
+  content_blocking_rules: CbRule[],
+  filters_used: string[],
+}
+
+export class FilterSet {
+    constructor(debug: boolean);
+    addFilter(filter: string, opts?: ParseOptions): null;
+    addFilters(rules: string[], opts?: ParseOptions): FilterListMetadata;
+    intoContentBlocking(): ContentBlockingConversionResult | undefined;
+}
+
+export const FilterFormat: {
+    HOSTS: string;
+    STANDARD: string;
+};
+
+export const RuleTypes: {
```

Comment:
I guess it should be an enum or union type. I do not know which will be compatible with the Rust library. At least it is enum here: https://docs.rs/adblock/latest/adblock/lists/enum.RuleTypes.html

For now, it is an object with all required fields and it can not be used to parse only some types of rules 

https://github.com/brave/adblock-rust/blob/dc2a4fa8bdda4e40f56deef63d4d39c0c925cce2/js/index.d.ts#L103

Reply:
hmm, after some reading ([1], [2]), my understanding is that `enum`s _shouldn't_ be declared in a `.d.ts` file.

This seems correct to me because it's describing the shape of `(await import('adblock-rs')).RuleTypes`, which basically behaves like a struct with those 3 fields.

Perhaps there is an issue with using `rule_types: typeof RuleTypes` in `ParseOptions` instead? I think it needs to be "field of" rather than `typeof`, but I don't know how to express that.

[1]: https://lukasbehal.com/2017-05-22-enums-in-declaration-files/
[2]: https://www.reddit.com/r/typescript/comments/dp7zd5/enum_or_enumlike_type_in_a_declaration_file/

Reply:
Maybe you can use `keyof typeof RyleTypes` or union type like here:
https://github.com/brave/adblock-rust/pull/343#discussion_r1530839332